tools/syz-declextract/syz-declextract.cpp

```diff
@@ -64,6 +106,230 @@ bool beginsWith(const std::string_view &str, const std::string_view begin) {
+}
+
+const std::string makeArray(const std::string &type, const size_t len = 0) {
+  if (len == 1) {
+    return type;
+  }
+  if (len) {
+    return "array[" + type + ", " + std::to_string(len) + "]";
+  }
+  return "array[" + type + "]";
+}
+
+const std::string makeArray(const std::string &type, const size_t min, const size_t max) {
+  if (min == max) {
+    return makeArray(type, min);
+  }
+  return "array[" + type + ", " + std::to_string(min) + ":" + std::to_string(max) + "]";
+}
+
+const std::string makePtr(const std::string &dir, const std::string &type, bool isOpt = false) {
+  if (type == "void") {
+    return makePtr(dir, makeArray("int8"));
+  }
+  std::string ptr = "ptr[" + dir + ", " + type;
+  if (isOpt) {
+    return ptr + ", opt]";
+  }
+  return ptr + "]";
+}
+
+const std::string makeConst(const std::string &type, const std::string &val = "0") {
+  if (type.empty()) {
+    return "const[" + val + "]";
+  }
+  if (type == "void") {
+    return "const[" + val + ", intptr]";
+  }
+  return "const[" + val + ", " + type + "]";
+}
+
+const std::string getSyzType(const std::string &ctype, const bool isSyscallParam = false) {
+  if (!isSyscallParam && (contains(ctype, "long long") || contains(ctype, "64"))) {
+    return contains(ctype, "be") ? "int64be" : "int64";
+  }
+  if (contains(ctype, "16") || contains(ctype, "short")) {
+    return contains(ctype, "be") ? "int16be" : "int16";
+  }
+  if (contains(ctype, "8")) {
+    return contains(ctype, "be") ? "int8be" : "int8";
+  }
+  if (contains(ctype, "32") || contains(ctype, "int")) {
+    return contains(ctype, "be") ? "int32be" : "int32";
+  }
+  if (contains(ctype, "void")) {
+    return "void";
+  }
+  if (contains(ctype, "char")) {
+    return "int8";
+  }
+  return "intptr";
+}
+
+class RecordExtractor {
+private:
+  const SourceManager *const SM;
+  std::vector<std::string> includes;
+  std::vector<std::string> flags;
+  std::unordered_map<std::string, size_t> visitedRecord; // Maps name to index in extractedRecords
+  std::vector<SyzRecordDecl> extractedRecords;
+
+  unsigned int getCountedBy(const FieldDecl *const &field) {
+    return field->getType()->isCountAttributedType()
+               ? llvm::dyn_cast<FieldDecl>(
+                     field->getType()->getAs<CountAttributedType>()->getCountExpr()->getReferencedDeclOfCallee())
+                     ->getFieldIndex()
+               : UINT_MAX;
+  }
+
+  bool isFieldVarlen(const QualType &fieldType) {
+    return fieldType->isIncompleteArrayType() ||
+           (fieldType->isConstantArrayType() && llvm::dyn_cast<ConstantArrayType>(fieldType)->getSize().isZero());
+  }
+
+  std::string getStructAttr(const RecordDecl *const recordDecl, ASTContext *context) {
+    if (recordDecl->isStruct()) {
+      for (const auto &item : recordDecl->getAttrs()) {
+        if (item->getKind() == clang::attr::Aligned) {
+          return "align[" + std::to_string(llvm::dyn_cast<AlignedAttr>(item)->getAlignment(*context)) + "]";
```

Comment:
Is this alignment in bytes or it bits? In the descriptions we need bytes. I suspect it may be wrong.